core/src/main/scala/kafka/server/ReplicaAlterDirThread.scala

```diff
@@ -0,0 +1,238 @@
+                                clientId = name,
+                                sourceBroker = sourceBroker,
+                                fetchBackOffMs = brokerConfig.replicaFetchBackoffMs,
+                                isInterruptible = false,
+                                includeLogTruncation = true) {
+
+  type REQ = FetchRequest
+  type PD = PartitionData
+
+  private val replicaId = brokerConfig.brokerId
+  private val maxBytes = brokerConfig.replicaFetchResponseMaxBytes
+  private val fetchSize = brokerConfig.replicaFetchMaxBytes
+
+  def fetch(fetchRequest: FetchRequest): Seq[(TopicPartition, PartitionData)] = {
+    var partitionData: Seq[(TopicPartition, FetchResponse.PartitionData)] = null
+    val request = fetchRequest.underlying.build()
+
+    def processResponseCallback(responsePartitionData: Seq[(TopicPartition, FetchPartitionData)]) {
+      partitionData = responsePartitionData.map { case (tp, data) =>
+        val abortedTransactions = data.abortedTransactions.map(_.asJava).orNull
+        val lastStableOffset = data.lastStableOffset.getOrElse(FetchResponse.INVALID_LAST_STABLE_OFFSET)
+        tp -> new FetchResponse.PartitionData(data.error, data.highWatermark, lastStableOffset,
+          data.logStartOffset, abortedTransactions, data.records)
+      }
+    }
+
+    replicaMgr.fetchMessages(
+      0L, // timeout is 0 so that the callback will be executed immediately
+      Request.FutureLocalReplicaId,
+      request.minBytes,
+      request.maxBytes,
+      request.version <= 2,
+      request.fetchData.asScala.toSeq,
+      UnboundedQuota,
+      processResponseCallback,
+      request.isolationLevel)
+
+    if (partitionData == null)
+      throw new IllegalStateException(s"Failed to fetch data for partitions ${request.fetchData.keySet().toArray.mkString(",")}")
+
+    partitionData.map { case (key, value) =>
+      key -> new PartitionData(value)
+    }
+  }
+
+  // process fetched data
+  def processPartitionData(topicPartition: TopicPartition, fetchOffset: Long, partitionData: PartitionData) {
+    val futureReplica = replicaMgr.getReplicaOrException(topicPartition, Request.FutureLocalReplicaId)
+    val partition = replicaMgr.getPartition(topicPartition).get
+    val records = partitionData.toRecords
+
+    if (fetchOffset != futureReplica.logEndOffset.messageOffset)
+      throw new RuntimeException("Offset mismatch for the future replica %s: fetched offset = %d, log end offset = %d.".format(
+        topicPartition, fetchOffset, futureReplica.logEndOffset.messageOffset))
+
+    // Append the leader's messages to the log
+    partition.appendRecordsToFutureReplica(records)
+    futureReplica.highWatermark = new LogOffsetMetadata(partitionData.highWatermark)
+    futureReplica.maybeIncrementLogStartOffset(partitionData.logStartOffset)
+
+    if (replicaMgr.getPartition(topicPartition).get.maybeDeleteAndSwapFutureReplica())
+      removePartitions(Set(topicPartition))
```

Comment:
I understand that ideally we want to shutdown idle threads. My concern with this approach is that it will let ReplicaAlterDirThread depend on ReplicaAlterDirManager which introduce circular dependency and make it easier to have deadlock in the future. 

The only drawback of the current approach is that, there may be idle ReplicaAlterDirThread until the broker receives the next LeaderAndIsrRequest. During this period this ReplicaAlterDirThread will call doWork() once every 1 second by default, which doesn't seem like a problem. Thus I think the little performance overhead is better than complicating the Kafka implementation. What do you think?

Reply:
Could we just add a scheduler thread that calls replicaAlterDirManager.shutdownIdleFetcherThreads() periodically?

Reply:
Good point. I have updated the patch to call replicaAlterDirManager.shutdownIdleFetcherThreads() every 2.5 seconds.